azavea/temperate
# Lock out expired users

## Overview

Directs users with expired subscriptions (of all types) to the expired page.

### Demo

![screen shot 2018-03-15 at 3 10 44 pm](https://user-images.githubusercontent.com/10568752/37485873-193c1890-2863-11e8-8959-f62e3bec8efc.png)


### Notes

As you will see, all this work happens front-end. One big downside is we can't test easily (would be ideal to test in Django).

I experimented with the server checking for errors (with a custom permissions class) and the front end handling the error (as suggested in the issue), and here's my takeaway.
- We would need to overhaul our use of `http` in Angular services and upgrade to[ `HttpClient`](https://angular.io/guide/http) which is[ new as of 4.3](https://github.com/angular/angular/commit/37797e2) in order to use Angular's [`Interceptor`](https://alligator.io/angular/httpclient-interceptors/) feature. This feature looks ideal for catching and handling request errors early -- earlier than route guards for sure. I don't know how much work it would entail to overhaul `http`, I stopped short there. I think this is the "right" way to do it ultimately, but can be deferred with this working solution with tradeoffs. `http` is on slow deprecation, and if we continue to work on Temperate past April 5 we will probably want to migrate to `HttpClient` anyway.

Mike brought up the UX -- if the server errors, work (in a rare case) would be lost and the user kicked out rather abruptly. Rather, by using a route guard, all work is saved and navigating to a new route kicks you to the expiration page.

ALSO, you will see the modal does not use the `ModalTemplateComponent`. This is because this modal is configured differently enough that I followed this [ngx-bootstrap example](https://valor-software.com/ngx-bootstrap/old/2.0.1/#/modals#directive-child) instead. Open to thoughts

ALSO x2 the use for the download csv component looks a little odd because it needed to be a button in once place, plain text in another.

## Testing Instructions
- In the django admin, change your org's subscription expiration date and time to a minute or two in the future. Make sure you can navigate around the app normally, no issues.
- While still valid, try hitting the `localhost:4210/expired` and see that you can't access that page
- Once expired, you should be kicked over to the `/expired` page on all internal app pages except the footer links and `/subscription`

Closes #602
